src/structure.js

```diff
@@ -53,10 +55,32 @@ function analyzeType(value) {
   };
 }
 
-function isa(Child, Ancestor) {
+export function isa(Child, Ancestor) {
   return Child === Ancestor || Child.prototype instanceof Ancestor;
 }
 
+export function collapseState(tree, value) {
+  return new $(tree)
+      .flatMap(node => {
```

Comment:
So we're converting all simple values into leaf nodes? Effectively truncating their children, right?

Reply:
Above comment was to make sure I understand correctly mainly for my understanding

I think that a `flatMap` is perfect here and now I know what you were talking about. The incredible thing about this approach is that even though the entire tree is theoretically part of the intermediate representation, the parts that aren't part of the final representation will never be computed because of laziness.

Reply:
> So we're converting all simple values into leaf nodes? Effectively truncating their children, right?

Yes.